### PR TITLE
Improve S3 Bitstream Storage to Lazy download object from S3

### PR DESCRIPTION
**Pull request by abollini**
Ported to 7.x in #9484 

## Description
The `S3BitStoreService#get` has been updated to return a custom InputStream able to lazy download the underlying S3 object when used. In this way we reduce the amount of space required on the local storage during download to a configurable value (default 5Mb).

> Please note that this PR contains also an unrelated commit 
> https://github.com/DSpace/DSpace/pull/9477/commits/2fd868095cf484179c7762a1af131c8083c2ca7e
> that is needed to allow to use a recent version of Eclipse to build DSpace. Without this commit eclipse complains about [The package org.w3c.dom is accessible from more than one module: <unnamed>, java.xml](https://stackoverflow.com/questions/57286825/the-package-org-w3c-dom-is-accessible-from-more-than-one-module-unnamed-java)

## Instructions for Reviewers
The `testBitstreamPutAndGetWithAlreadyPresentBucket` in dspace-api/src/test/java/org/dspace/storage/bitstore/S3BitStoreServiceIT.java has been updated to verify that get and put still work properly and are consistent for different content size.

You should checkout the PR and try to download one or more files of different size, monitor the temp folder to verify that no files named s3-copy-... are left behind and that also downloading a large file each temp file is not larger than 5Mb.

You can eventually customize the size of the chunk in the `bitstore.xml` spring file setting the property `bufferSize` in the `s3Store` bean definition.

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [n/a] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [X] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [n/a] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 1]
```diff
@@ -622,4 +617,68 @@ public boolean isRegisteredBitstream(String internalId) {
         return internalId.startsWith(REGISTERED_FLAG);
     }
 
+    public void setBufferSize(long bufferSize) {
+        this.bufferSize = bufferSize;
+    }
+
+    public class S3LazyInputStream extends InputStream {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:624]: Can we add JavaDocs to describe this imbedded class?

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 2]
```diff
@@ -622,4 +617,68 @@ public boolean isRegisteredBitstream(String internalId) {
...
+			int byteRead = currPos < endOfChunk ? currentChunkStream.read() : -1;
+            if (byteRead != -1) {
+                currPos++;
+            } else {
+                currentChunkStream.close();
+            }
+            return byteRead;
+        }
+
+		private void downloadChunk() throws IOException, FileNotFoundException {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:656]: Alignment of this method is off.  Can we also add a basic description to this class to describe what it's doing?  This seems to be the main method.

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 3]
```diff
@@ -622,4 +617,68 @@ public boolean isRegisteredBitstream(String internalId) {
...
+        }
+
+        @Override
+        public int read() throws IOException {
+            if (currPos == endOfChunk && currPos < fileSize) {
+                currentChunkStream.close();
+                downloadChunk();
+            }
+
+			int byteRead = currPos < endOfChunk ? currentChunkStream.read() : -1;
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:647]: Code alignment it odd here. It's indented more than normal

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java [hunk 4]
```diff
@@ -258,20 +266,7 @@ public InputStream get(Bitstream bitstream) throws IOException {
...
-        try {
-            File tempFile = File.createTempFile("s3-disk-copy-" + UUID.randomUUID(), "temp");
-            tempFile.deleteOnExit();
-
-            GetObjectRequest getObjectRequest = new GetObjectRequest(bucketName, key);
-
-            Download download = tm.download(getObjectRequest, tempFile);
-            download.waitForCompletion();
-
-            return new DeleteOnCloseFileInputStream(tempFile);
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:270]: Since you've replaced the `DeleteOnCloseFileInputStream`, should we *delete* that class? It looks like it is no longer used.

**Review by tdonohue — COMMENTED** [review]
@abollini :  I haven't tested this yet, but the code looks good overall.  A few minor code suggestions inline below.

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java (hunk 4 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:270]: it is still used by the inner class, see https://github.com/DSpace/DSpace/pull/9477/files#diff-02dc8d6ccf99e2499bb006b204a26eb968f83223394b5b50a386d7be887ba8adR682

File: dspace-api/pom.xml [hunk 5]
```diff
@@ -609,6 +609,12 @@
         <dependency>
             <groupId>org.apache.tika</groupId>
             <artifactId>tika-parsers-standard-package</artifactId>
+            <exclusions>
+                <exclusion>
+                    <groupId>xml-apis</groupId>
+                    <artifactId>xml-apis</artifactId>
+                </exclusion>
+            </exclusions>
```
Comment (pnbecker) [inline dspace-api/pom.xml:617]: Why do we need to exclude this? How is this related to the issue this PR is solving?

File: dspace-server-webapp/pom.xml [hunk 6]
```diff
@@ -499,6 +499,12 @@
         <dependency>
             <groupId>org.dspace</groupId>
             <artifactId>dspace-sword</artifactId>
+            <exclusions>
+                <exclusion>
+                    <groupId>xml-apis</groupId>
+                    <artifactId>xml-apis</artifactId>
+                </exclusion>
+            </exclusions>
```
Comment (pnbecker) [inline dspace-server-webapp/pom.xml:507]: See comment at the other pom.xml

**Comment by pnbecker** [comment]
I will have to look deeper into this. In a first quick test I was not able to download files anymore. I got a White label error page instead.

File: dspace-server-webapp/pom.xml (hunk 6 above)
Comment (tdonohue) [reply dspace-server-webapp/pom.xml:507]: See above comment...this change seems to be related to an issue with using Eclipse + DSpace.  It is unrelated to the S3 changes but I expect @abollini made the change to allow his Eclipse IDE to work better in developing this fix.

**Comment by abollini** [comment]
@tdonohue rebase performed, I have also removed my changes for the xml-apis and verified that the already merged solution still work in eclipse ;)

**Comment by tdonohue** [comment]
@pnbecker or @kshepherd : Could one of you test this PR (or the 7.x version in #9484 ) to verify it's now working?  I'd like to get this merged, but I need to locate an Amazon S3 user to test this out.  The code looks reasonable to me though.

**Review by pnbecker — APPROVED** [review]
Code looks great, thank you. I tested the version rebased to DSpace 7 and it works well. Thank you @abollini!

**Review by tdonohue — APPROVED** [review]
👍 Thanks @abollini !  The code looks good to me.

**Comment by tdonohue** [comment]
Ported to 7.x in #9484
